File: Recursion/PalindromString/PalindromString.java
import java.util.*;
import java.io.*;

public class PalindromString {
    public static void main(String [] args  ){
        String str ="geeks";
        ArrayList<String> out= new ArrayList<>();
        ArrayList<ArrayList<String>> res = new ArrayList<ArrayList<String>>();
        palindromPartHelper(str,0,"",res);
        System.out.println(res);
        
    }
    public static void palindromPartHelper(String str,int index,String out,ArrayList<ArrayList<String>> res){
        if (index == str.length()){
            String [] arr = out.split(" ");
            List<String> list = Arrays.asList(arr);
            list=list.subList(1,list.size());
            res.add(new ArrayList<>(list));
        }
        for (int i=index;i<str.length();i++){
            if (isPalindrom(str.substring(index,i+1))){
                palindromPartHelper(str,i+1,out+" "+str.substring(index,i+1),res);
                
            }
            
        }
    }
    public static boolean isPalindrom(String str){
        int len = str.length();
        int i=0;
        int j = len-1;
        while(i<=j){
            if (str.charAt(i)!=str.charAt(j)){
                return false;
            }
            i++;
            j--;
                    
        }
        return true;
    }
}

File: Recursion/PalindromString/palindromString.go
package main

import "fmt"

func main(){
	var str="geeks"
	createPalindromstring(str,0,"")
}
func createPalindromstring(str string,index int,out string){
	if index==len(str){
		fmt.Println(out)
	}
	for i:=index;i<len(str);i++{
		if isPalindrom(str[index:i+1]){
			createPalindromstring(str,i+1,out+" "+str[index:i+1])
		}
	}
}
func isPalindrom(str string)bool{
	if len(str)<2{
		return true
	}
	for i,j:=0,len(str)-1;i<=j;i,j=i+1,j-1{
		if str[i]!=str[j]{
			return false
		}

	}
	return true
}
